File: src/utils/sourcePages.ts
/** 族谱原文按页拆分/合并（与 backend/source_pages.py 规则一致） */

export type SourcePageSlice = {
  page: number
  text: string
}

const PAGE_HEADER_RE = /^===== 第\s*(\d+)\s*(?:页|张)(?:\s*·.*?)?\s*=====\s*$/gm

export function splitPagedText(text: string): SourcePageSlice[] {
  const raw = (text || '').trim()
  if (!raw) return []

  const headers: { page: number; start: number; end: number }[] = []
  let match: RegExpExecArray | null
  const re = new RegExp(PAGE_HEADER_RE.source, 'gm')
  while ((match = re.exec(raw)) !== null) {
    headers.push({
      page: Number(match[1]),
      start: match.index,
      end: match.index + match[0].length,
    })
  }

  if (!headers.length) return [{ page: 1, text: raw }]

  const out: SourcePageSlice[] = []
  for (let i = 0; i < headers.length; i += 1) {
    const h = headers[i]
    const chunkStart = h.end
    const chunkEnd = i + 1 < headers.length ? headers[i + 1].start : raw.length
    out.push({
      page: h.page,
      text: raw.slice(chunkStart, chunkEnd).trim(),
    })
  }
  return out
}

export function mergePagedText(pages: SourcePageSlice[], unit: '页' | '张' = '页'): string {
  const parts: string[] = []
  for (const item of pages) {
    const text = (item.text || '').trim()
    if (!text) continue
    parts.push(`===== 第 ${item.page} ${unit} =====\n${text}`)
  }
  return parts.join('\n\n')
}

export function getPageText(text: string, page: number): string {
  return splitPagedText(text).find((p) => p.page === page)?.text || ''
}

export function setPageText(text: string, page: number, newText: string): string {
  let pages = splitPagedText(text)
  if (!pages.length) pages = [{ page: 1, text: '' }]
  const idx = pages.findIndex((p) => p.page === page)
  const next = (newText || '').trim()
  if (idx >= 0) pages[idx] = { ...pages[idx], text: next }
  else {
    pages = [...pages, { page, text: next }].sort((a, b) => a.page - b.page)
  }
  return mergePagedText(pages)
}

export function resolvePageCount(ocrText: string, imageCount: number): number {
  const fromText = splitPagedText(ocrText).length
  return Math.max(fromText, imageCount, 1)
}

export function buildPageList(
  ocrText: string,
  relationText: string,
  imagePaths: string[] = [],
  imagePreviews: string[] = [],
): Array<{
  page: number
  ocrText: string
  relationText: string
  imagePath?: string
  imagePreview?: string
}> {
  const count = resolvePageCount(ocrText, Math.max(imagePaths.length, imagePreviews.length))
  const ocrPages = splitPagedText(ocrText)
  const relPages = splitPagedText(relationText)
  const ocrMap = new Map(ocrPages.map((p) => [p.page, p.text]))
  const relMap = new Map(relPages.map((p) => [p.page, p.text]))

  const pages: Array<{
    page: number
    ocrText: string
    relationText: string
    imagePath?: string
    imagePreview?: string
  }> = []

  for (let i = 0; i < count; i += 1) {
    const page = ocrPages[i]?.page ?? relPages[i]?.page ?? i + 1
    pages.push({
      page,
      ocrText: ocrMap.get(page) || ocrPages[i]?.text || '',
      relationText: relMap.get(page) || relPages[i]?.text || '',
      imagePath: imagePaths[i],
      imagePreview: imagePreviews[i],
    })
  }
  return pages
}

File: backend/source_pages.py
"""族谱原文按页拆分/合并（版本一 OCR、版本二关系描述共用页码）。"""

from __future__ import annotations

import re
from typing import Any

PAGE_HEADER_RE = re.compile(
    r"^===== 第\s*(\d+)\s*(?:页|张)(?:\s*·.*?)?\s*=====\s*$",
    re.MULTILINE,
)


def split_paged_text(text: str) -> list[dict[str, Any]]:
    """将带页码分隔符的全文拆成 [{page, text}, ...]；无分隔符时视为第 1 页。"""
    raw = (text or "").strip()
    if not raw:
        return []

    matches = list(PAGE_HEADER_RE.finditer(raw))
    if not matches:
        return [{"page": 1, "text": raw}]

    out: list[dict[str, Any]] = []
    for idx, match in enumerate(matches):
        page = int(match.group(1))
        start = match.end()
        end = matches[idx + 1].start() if idx + 1 < len(matches) else len(raw)
        chunk = raw[start:end].strip()
        out.append({"page": page, "text": chunk})
    return out


def merge_paged_text(pages: list[dict[str, Any]], *, unit: str = "页") -> str:
    """合并分页文本，页码分隔符与 OCR 批量识别一致。"""
    parts: list[str] = []
    for item in pages:
        page = int(item.get("page") or len(parts) + 1)
        text = (item.get("text") or "").strip()
        if not text:
            continue
        parts.append(f"===== 第 {page} {unit} =====\n{text}")
    return "\n\n".join(parts)


def get_page_text(text: str, page: int) -> str:
    for item in split_paged_text(text):
        if int(item.get("page") or 0) == page:
            return (item.get("text") or "").strip()
    return ""


def set_page_text(text: str, page: int, new_text: str) -> str:
    pages = split_paged_text(text)
    if not pages:
        pages = [{"page": 1, "text": ""}]
    found = False
    for item in pages:
        if int(item.get("page") or 0) == page:
            item["text"] = (new_text or "").strip()
            found = True
            break
    if not found:
        pages.append({"page": page, "text": (new_text or "").strip()})
        pages.sort(key=lambda x: int(x.get("page") or 0))
    return merge_paged_text(pages)


def page_count(text: str) -> int:
    pages = split_paged_text(text)
    return max((int(p.get("page") or 0) for p in pages), default=1) if pages else 1
